### PR TITLE
Add batch-lines support to Redis destination

### PR DESCRIPTION
**Pull request by OverOrion**
This PR adds `batch-lines()` support for the `redis` destination driver which increases performance.
Performance improvement is significant (8665U, 32 GiB RAM): [1]

| 8 workers, no batch-lines() option | 8 workers, batch-lines(100) |
|------------------------------------|-----------------------------|
| ~ 80k/s                            | ~ 240k/s (peak 450k/s)      |

The batching is based on the [pipelining feature of Redis](https://github.com/redis/hiredis#pipelining). We have to count the sent (appended) messages as we must retrieve as many as we sent.

I also tested that when the Redis server goes offline it will reconnect.

[1] I used the following configuration:
```
source s_local {
        network(
        port(4444)
        max-connections(100)
        log-iw-size(100000)
        log-fetch-limit(1000)
        );
};
destination d_redis {
    redis(
        host("localhost")
        port(6379)
        command("HINCRBY", "hosts", "$HOST", "1")
        workers(8)
        log-fifo-size(100000)
        batch-lines(100)
    );
};

log {
        source(s_local);
        destination(d_redis);
        flags(flow-control);
};
```
And sent messages via loggen:
`./loggen --permanent --rate=10000000 --active-connections 10 -i localhost 4444`

Signed-off-by: Parrag Szilárd <szilard.parrag@gmail.com>

### CODE REVIEW

**Review by Barkodcz — DISMISSED** [review]
Nice, thank you!

File: modules/redis/redis-worker.c [hunk 1]
```diff
@@ -277,17 +352,19 @@ redis_worker_connect(LogThreadedDestWorker *s)
...
 
-  log_threaded_dest_worker_init_instance(&self->super, owner, worker_index);
+  log_threaded_dest_worker_init_instance(&self->super, o, worker_index);
 
   self->super.thread_init = redis_worker_thread_init;
   self->super.thread_deinit = redis_worker_thread_deinit;
   self->super.connect = redis_worker_connect;
   self->super.disconnect = redis_worker_disconnect;
-  self->super.insert = redis_worker_insert;
+  self->super.insert = owner->super.batch_lines > 0 ? redis_worker_insert_batch : redis_worker_insert;
```
Comment (MrAnno) [inline modules/redis/redis-worker.c:366]: The `batch-lines()` option is missing from the `redis()` modules.
Please add `threaded_dest_driver_batch_option` to the grammar.

**Review by MrAnno — DISMISSED** [review]
Please add a news file entry about the new options: `batch-lines()` and `batch-timeout()`.

File: modules/redis/redis-worker.c (hunk 1 above)
Comment (MrAnno) [reply modules/redis/redis-worker.c:366]: Is it guaranteed that this constructor runs only after batch_lines is set (after parsing and setting all config options)?

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: modules/redis/redis-worker.c [hunk 2]
```diff
@@ -229,12 +300,18 @@ redis_worker_connect(LogThreadedDestWorker *s)
...
-    return TRUE;
-
-  if (self->c)
-    redisFree(self->c);
+    {
+      return TRUE;
+    }
+  else if(self->c)
+    {
+      redisReconnect(self->c);
```
Comment (Kokan) [inline modules/redis/redis-worker.c:308]: Why do we need this change ? Preferably move it into a different commit and explain in the commit message the reason behind this change.

File: news/feature-3745.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+I used the following configuration to enable batching:
+```
+destination d_redis {
+    redis(
+        host("localhost")
+        port(6379)
+        command("HINCRBY", "hosts", "$HOST", "1")
+        workers(8)
+        log-fifo-size(100000)
+        batch-lines(100)
```
Comment (Kokan) [inline news/feature-3745.md:14]: This also adds the `batch-timeout` option. Could you please also provide some information how you did that benchmark (configuration, setup).

File: modules/redis/redis-worker.h [hunk 4]
```diff
@@ -40,6 +40,9 @@ typedef struct _RedisDestWorker
   gchar **argv;
   size_t *argvlen;
 
+  /*Batch mode*/
+  size_t num_of_messages_waiting_for_reply;
```
[the remark this answers is not in the document]
Comment (OverOrion) [reply modules/redis/redis-worker.h:44]: In my understanding the `batch_size` will be distributed among the workers and that is why each worker has to know how many messages it sent so it can request as many replies.

[the remark this answers is not in the document]
Comment (Kokan) [reply modules/redis/redis-worker.h:44]: The `batch_size` is the property of `LogThrreadedDestWriter` thus it is unique to the specific worker. Check when and how it is manipulated in the `lib/logthrdest/logthrdestdrv.c` file.

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by Kokan — APPROVED** [review]

**Review by MrAnno — APPROVED** [review]
